File: packages/graph-view/src/wall.ts
/**
 * PHASE 1 — THE WALL (graph-view-wall/1.0.0).
 *
 * The minimal, clean, versioned, typed face this cell presents to its neighbors
 * (SEAM-MAP V4: the hub serves a canonical envelope; V5: the editor joins via the
 * shared-ID event bus) — promoted OVER the cell's diagnostic pins, never replacing
 * them (WALL-CONVENTIONS rules 1–3). The wall wraps runGraphView; it adds NO new
 * pipeline behavior, only:
 *   - a construction-time schema-PIN assert (rule 4) — the wall carries its own
 *     copy of the pinned (schemaVersion, schemaHash) pair and refuses loudly to
 *     stand on a drifted canonical schema (failure-class=schema-pin-mismatch);
 *   - the pins accessor {probeCatalog, dump, history, tap} delegating to the
 *     cell's existing introspection quartet (rule 3 — the 10 Fitbit pins stay
 *     reachable under the 2-pin charger);
 *   - envelope honesty: a served graph the pins prove was rejected whole
 *     (ingest.envelope.version ok:false) is a wall REJECTION with a named
 *     failure class and the pins attached — never a silently-empty render;
 *   - four wall.* leads (cataloged; the catalog only grows) so the wall's own
 *     decisions are on the same probe stream as everything else.
 *
 * The bus argument is THIS cell's GraphEventBus interface exactly as it stands
 * today ({type:'select', nodeId, source} + optional hover — see eventBus.ts,
 * quoted verbatim in MEMBRANE-SPEC.md). Phase 2 builds the T4↔T5 shape-mapping
 * adapter (SEAM-MAP V5); the wall does NOT adapt to the editor's shape.
 *
 * SUB200 restructure: the wall's carried identity, contracts, and error class
 * live in wallTypes.ts, re-exported here — importers of "./wall" are unchanged.
 */

import { runGraphView, type GraphViewCell } from "./cell";
import { ProbeBus } from "./probeBus";
import { KNOWN_PROBE_IDS, probeCatalog } from "./probeCatalog";
import type { GraphEventBusWithHover } from "./eventBus";
import { checkPin, PINNED_SCHEMA_VERSION } from "../../schema/gen/pin";
import { SCHEMA_VERSION } from "./schema";
import {
  GraphViewWallError, WALL_SCHEMA_HASH, WALL_SCHEMA_VERSION, WALL_VERSION,
  type WallFailureClass, type WallOptions, type WallPins,
} from "./wallTypes";

export {
  GraphViewWallError, WALL_SCHEMA_HASH, WALL_SCHEMA_VERSION, WALL_VERSION,
} from "./wallTypes";
export type { WallFailureClass, WallOptions, WallPins } from "./wallTypes";

const STAGE = "W"; // the wall is not a pipeline stage; its leads sit beside S0–S7

/** The face: the cell handle (unchanged — the wall promotes it, additive-only)
 *  plus the pins quartet. Nothing else leaks through (rule 2). */
export interface GraphViewWall {
  cell: GraphViewCell;
  pins: WallPins;
}

export async function createGraphViewWall(
  schemaJson: unknown,
  eventBus: GraphEventBusWithHover,
  opts: WallOptions = {},
): Promise<GraphViewWall> {
  // The wall owns the probe bus so its own leads AND the run's leads share one
  // stream — and so pins survive any rejection raised below.
  const probeBus = new ProbeBus(KNOWN_PROBE_IDS);

  const constructCause = probeBus.emit({
    probeId: "wall.construct", stage: STAGE, kind: "call",
    payload: {
      wallVersion: WALL_VERSION,
      schemaVersion: WALL_SCHEMA_VERSION,
      schemaHash: WALL_SCHEMA_HASH,
      capConfigOverridden: opts.capConfig !== undefined,
    },
    causeId: null,
  });

  // ── PIN assert at construction (rule 4): the wall's carried pin vs the
  // canonical gen/pin.ts constants, plus the generated-artifact cross-check
  // (gen/graph-schema.ts SCHEMA_VERSION vs gen/pin.ts PINNED_SCHEMA_VERSION —
  // a partially regenerated gen/ folder is ALSO drift). Refuse loudly on drift.
  let pinReason: string | null = null;
  try {
    checkPin(WALL_SCHEMA_VERSION, WALL_SCHEMA_HASH);
    if ((SCHEMA_VERSION as string) !== (PINNED_SCHEMA_VERSION as string)) {
      throw new Error(
        `failure-class=schema-pin-mismatch: gen/graph-schema.ts SCHEMA_VERSION ${JSON.stringify(SCHEMA_VERSION)}` +
        ` != gen/pin.ts PINNED_SCHEMA_VERSION ${JSON.stringify(PINNED_SCHEMA_VERSION)} — partially regenerated schema package`,
      );
    }
  } catch (e) {
    pinReason = e instanceof Error ? e.message : String(e);
  }
  const pinAssertCause = probeBus.emit({
    probeId: "wall.pin.assert", stage: STAGE, kind: "decision",
    payload: {
      ok: pinReason === null,
      carriedVersion: WALL_SCHEMA_VERSION,
      pinnedVersion: PINNED_SCHEMA_VERSION,
      carriedHash: WALL_SCHEMA_HASH,
      ...(pinReason !== null ? { reason: pinReason } : {}),
    },
    causeId: constructCause,
  });
  if (pinReason !== null) {
    probeBus.emit({
      probeId: "wall.face.reject", stage: STAGE, kind: "error",
      payload: { failureClass: "schema-pin-mismatch", reason: pinReason },
      causeId: pinAssertCause,
    });
    // No cell exists yet; history/catalog stay reachable through the error's pins.
    throw new GraphViewWallError("schema-pin-mismatch", pinReason, busOnlyPins(probeBus));
  }

  // ── Run the cell EXACTLY as it stands (the wall adds nothing to S0–S7).
  let cell: GraphViewCell;
  try {
    cell = await runGraphView(schemaJson, eventBus, { capConfig: opts.capConfig, bus: probeBus });
  } catch (err) {
    // A cell gate violation (EdgeSetViolation / NodeSetViolation /
    // LeadPromotionViolation) crosses the wall UNCHANGED — probed, then rethrown.
    probeBus.emit({
      probeId: "wall.face.reject", stage: STAGE, kind: "error",
      payload: {
        failureClass: err instanceof Error ? err.name : "unknown",
        reason: err instanceof Error ? err.message : String(err),
      },
      causeId: pinAssertCause,
    });
    throw err;
  }

  const pins: WallPins = {
    probeCatalog: () => cell.probeCatalog(),
    dump: () => cell.dump(),
    history: () => cell.history(),
    tap: (probeId, fn) => cell.tap(probeId, fn),
  };

  // ── Envelope honesty: the wall declares ONLY what the pins prove was ingested.
  // A payload ingest rejected whole (ruling 3 + pin) must not surface as a
  // silently-empty-but-standing wall — it is a named rejection, pins attached.
  const envelopePin = probeBus.last("ingest.envelope.version");
  const env = envelopePin?.payload as { present: boolean; ok: boolean; value?: string } | undefined;
  if (env === undefined || !env.ok) {
    const failureClass: WallFailureClass = env?.present ? "schema-pin-mismatch" : "envelope-rejected";
    const reason = env === undefined
      ? "ingest.envelope.version pin missing from the stream — the wall cannot prove the envelope was checked"
      : env.present
        ? `served envelope schemaVersion ${JSON.stringify(env.value)} != pinned ${JSON.stringify(PINNED_SCHEMA_VERSION)} — rejected whole at ingest (assembly ruling 3 + schema pin)`
        : `served envelope has no schemaVersion — the canonical envelope {schemaVersion:${JSON.stringify(PINNED_SCHEMA_VERSION)},nodes,edges,leads} requires it (assembly ruling 3); rejected whole at ingest`;
    probeBus.emit({
      probeId: "wall.face.reject", stage: STAGE, kind: "error",
      payload: { failureClass, reason },
      causeId: envelopePin ? `${envelopePin.probeId}#${envelopePin.logicalClock}` : pinAssertCause,
    });
    throw new GraphViewWallError(failureClass, reason, pins);
  }

  // ── The declared face summary — one pinned lead the conformance test holds
  // against pin-level truth (declared may never diverge from probed).
  probeBus.emit({
    probeId: "wall.face.result", stage: STAGE, kind: "output",
    payload: {
      wallVersion: WALL_VERSION,
      nodes: cell.rfNodes.length,
      edges: cell.rfEdges.length,
      leads: cell.rfEdges.filter((e) => e.type === "leadEdge").length,
      engine: cell.engine.name,
      capMode: cell.cap.mode,
      bannerShown: cell.banner.bannerShown,
    },
    causeId: pinAssertCause,
  });

  return { cell, pins };
}

/** Pins for a wall that refused BEFORE a cell existed: the catalog is static and
 *  history/tap come from the wall's own bus; dump() honestly refuses — there is
 *  no cell state to dump, and fabricating one would be a fake pin. */
function busOnlyPins(probeBus: ProbeBus): WallPins {
  return {
    probeCatalog: () => probeCatalog(),
    dump: () => {
      throw new GraphViewWallError(
        "schema-pin-mismatch",
        "wall refused at construction — no cell stands, so there is no dump; read history()/tap() for the refusal stream",
      );
    },
    history: () => probeBus.history(),
    tap: (probeId, fn) => probeBus.tap(probeId, fn),
  };
}

File: packages/structure-extractor/wall.py
"""structure-extractor WALL — Phase 1 (WALL-CONVENTIONS.md; face frozen by SEAM-MAP.md).

The minimal, clean, versioned, typed face this cell presents to its neighbors,
promoted OVER the diagnostic pins — never replacing them:

    from wall import extract_wall, WALL_VERSION

    wall = extract_wall()
    env  = wall.extract(root)            # {schemaVersion:"v0", nodes, edges, leads,
                                         #  honestCeilings} — the canonical envelope
    wall.honestCeiling("python")         # the dock's declared ceiling (or typed refusal)
    wall.pins.probeCatalog()             # the full diagnostic quartet stays reachable:
    wall.pins.dump()                     #   probeCatalog / dump / history / tap
    wall.pins.history()                  #   ("10 Fitbit pins under the 2-pin charger")
    wall.pins.tap(id, fn)

Rules this module enforces (see MEMBRANE-SPEC.md for the one-page contract):
  * The wall asserts the schema PIN at construction AND before every extract():
    it refuses to stand on a drifted schema (failure class schema-pin-mismatch).
  * The face returns ONLY the canonical envelope + per-dock honest ceilings.
    t3/summary/decisions stay PINS (pins.dump()/history()) — the composed-graph
    gap analysis happens in Phase 3 over the whole organism, so the cell's
    internal t3 is diagnostic, not the wall face.
  * Honest ceiling propagates: an undeclared or unknown ceiling surfaces as
    such — the wall never fabricates a tier, a verdict, or green.
  * Every wall rejection is a NAMED failure class (FAILURE_CLASSES), and
    every wall decision is probed via catalogued extractor.wall.* leads.

V1 socket (capability wall -> extractor wall): `capability_fn` on extract() is
the formalized injection point; it defaults to the cell's local stub_capability.
KNOWN FRICTION, deliberately NOT fixed here (vessel work, not wall work):
CapabilityHandle.handle is never consumed by ExtractorCell._make_dock — the
python dock builds its own pyright backend from PipelineConfig.  Wiring Tree
2's real negotiated handle requires a Phase-2 adapter touching _make_dock.

SUB200 restructure: failure classes, the envelope guard, the pins accessor and
config merging live in extractor/wall_support.py; the schema-PIN assertion in
extractor/wall_pin.py; this module re-exports every one of them (facade —
`from wall import ...` is byte-for-byte the same surface).
"""
from __future__ import annotations

from pathlib import Path
from typing import Callable

from extractor import schema as cell_schema
from extractor.assemble import (FakedEdgeError, TierInflationError,  # noqa: F401
                                UnbackedGreenError)
from extractor.capability import CapabilityHandle, stub_capability  # noqa: F401 (face re-export)
from extractor.pipeline import ExtractorCell, PipelineConfig
from extractor.wall_pin import assert_schema_pin
from extractor.wall_support import (FAILURE_CLASSES, IdMismatchError,  # noqa: F401
                                    LeadInEdgesError, UnknownLanguageError,
                                    WallPins, WallRefusal,
                                    WallSchemaPinMismatch, _merge_config,
                                    validate_envelope)

WALL_VERSION = "structure-extractor-wall/1.0.0"

_STAGE = "WALL"


class ExtractorWall:
    """The Phase-1 face of tree3.structure-extractor (see module docstring)."""

    WALL_VERSION = WALL_VERSION

    def __init__(self, config: PipelineConfig | dict | None = None):
        self._base_config = _merge_config(config, None, None)
        self._cell = ExtractorCell(self._base_config)
        self.pins = WallPins(self)
        self._cell.bus.emit("extractor.wall.version", _STAGE, "state", {
            "wallVersion": WALL_VERSION,
            "pinnedSchemaVersion": cell_schema.PINNED_SCHEMA_VERSION,
            "pinnedSchemaHash": cell_schema.PINNED_SCHEMA_HASH})
        self._assert_schema_pin()

    # ---- schema PIN (wall convention rule 4; wall_support.assert_schema_pin)
    def _assert_schema_pin(self) -> None:
        assert_schema_pin(self._cell.bus)

    # ---- the face ------------------------------------------------------------
    def extract(self, root: str | Path,
                capability_fn: Callable[[str], CapabilityHandle] | None = None,
                out_dir: str | Path | None = None,
                config: PipelineConfig | dict | None = None) -> dict:
        """Run the cell over `root`, return the canonical Graph envelope
        {schemaVersion:"v0", nodes, edges, leads} + per-dock honestCeilings.

        `capability_fn` is the formalized V1 socket (defaults to the local
        stub_capability; Phase 2 plugs Tree 2's real function).  t3/summary
        do NOT cross the face — they stay pins (pins.dump()/history())."""
        base = config if config is not None else self._base_config
        cfg = _merge_config(base, capability_fn, out_dir)
        cell = self._cell
        cell.config = cfg
        self._assert_schema_pin()   # re-asserted before EVERY run, not just birth
        cell.bus.emit("extractor.wall.extract.call", _STAGE, "call", {
            "root": str(root),
            "capabilityFnInjected": capability_fn is not None,
            "outDir": str(cfg.out_dir) if cfg.out_dir is not None else None})
        result = cell.run(Path(root))
        envelope = {
            "schemaVersion": result["schemaVersion"],
            "nodes": result["nodes"],
            "edges": result["edges"],
            "leads": result["leads"],
            "honestCeilings": result["honestCeilings"],
        }
        try:
            validate_envelope(envelope)
        except WallRefusal as exc:
            cell.bus.emit("extractor.wall.refusal", _STAGE, "error",
                          {"failureClass": exc.failure_class, "detail": str(exc)})
            raise
        cell.bus.emit("extractor.wall.extract.return", _STAGE, "output", {
            "nodes": len(envelope["nodes"]), "edges": len(envelope["edges"]),
            "leads": len(envelope["leads"]),
            "docks": sorted(envelope["honestCeilings"])})
        return envelope

    def honestCeiling(self, lang: str) -> dict:
        """The dock's declared ceiling for `lang`, VERBATIM from the pin
        surface (== extractor.output.honest_ceiling.report perDock[lang]).

        * lang outside the Frozen-Schema set -> UnknownLanguageError (typed
          refusal, consistent with the capability wall's vocabulary: no depth
          tier CT/S/G/P exists for it and none is fabricated).
        * lang known but no dock ceiling declared in the last extract()
          (no run yet / no sources / no dock this round) -> an explicit
          undeclared surface {declared: False, tier: None, ...} — reduced
          surfaces as reduced, never green."""
        bus = self._cell.bus
        if lang not in cell_schema.LANGS:
            err = UnknownLanguageError(
                f"lang {lang!r} is outside the Frozen-Schema language set "
                f"{list(cell_schema.LANGS)}: no capability tier (CT/S/G/P) is "
                f"negotiated for it and the wall will not fabricate one "
                f"(unknown != green)")
            bus.emit("extractor.wall.honest_ceiling.request", _STAGE, "call",
                     {"lang": lang, "outcome": "refused-unknown-language"})
            bus.emit("extractor.wall.refusal", _STAGE, "error",
                     {"failureClass": err.failure_class, "detail": str(err)})
            raise err
        state = self._cell.dump()
        declared = {c["lang"]: c for c in state["honestCeilings"]}
        if lang in declared:
            bus.emit("extractor.wall.honest_ceiling.request", _STAGE, "call",
                     {"lang": lang, "outcome": "declared"})
            return declared[lang]
        sourcesets = state["sourcesets"]
        if not sourcesets:
            reason = "no extract() has run through this wall yet"
        elif lang not in sourcesets:
            reason = f"the last extract() ingested no {lang} sources"
        else:
            reason = (f"{lang} had sources but no T2 dock this round "
                      f"(Go/C/C++ are T1-only: structure nodes, zero resolved edges)")
        bus.emit("extractor.wall.honest_ceiling.request", _STAGE, "call",
                 {"lang": lang, "outcome": "undeclared"})
        return {
            "lang": lang,
            "declared": False,
            "tier": None,   # never fabricated: an unknown tier surfaces as unknown
            "resolves": [],
            "cannotResolve": ["everything — no dock ceiling was declared for this "
                              "language by the wall's last extract()"],
            "resolverGrade": "none (undeclared — an unknown ceiling is surfaced as "
                             "unknown, never fabricated)",
            "blindSpots": [reason],
        }


def extract_wall(config: PipelineConfig | dict | None = None) -> ExtractorWall:
    """Wall factory — the exported constructor neighbors use."""
    return ExtractorWall(config)


__all__ = [
    "WALL_VERSION", "ExtractorWall", "extract_wall", "WallPins",
    "PipelineConfig", "CapabilityHandle", "stub_capability",
    "validate_envelope", "FAILURE_CLASSES",
    "WallRefusal", "WallSchemaPinMismatch", "UnknownLanguageError",
    "LeadInEdgesError", "IdMismatchError", "TierInflationError", "FakedEdgeError",
    "UnbackedGreenError",
]
